Ryan 21003 beta (#2043)
OpenLiberty/blogs: posts/2022-02-17-beta-release-post.adoc

author_github: https://github.com/ryan-storey
seo-title: Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta - OpenLiberty.io
seo-description: Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
blog_description: Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> It's only been a few short months since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, and we're already looking forward as we collaborate on and begin to implement Jakarta EE 10 features.  Concurrency 3.0 marks our first Jakarta EE 10 beta feature, available in Open Liberty 22.0.0.3-beta.  Concurrency 3.0 introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.

Reviewer: SInce the 3.0 enhancements are detailed later, you might summarize them here to make the intro more concise, here and in the metadata descriptions. Also only need one space after a period

Although only a few short months have passed since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, we're already implementing Jakarta EE 10 features. Concurrency 3.0 marks our first Jakarta EE 10 beta feature, available in Open Liberty 22.0.0.3-beta. Concurrency 3.0 introduces various enhancements, including new resource definition annotations, container-managed context-aware completion stages, and more.

> It's only been a few short months since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, and we're already looking forward as we collaborate on and begin to implement Jakarta EE 10 features.  Concurrency 3.0 marks our first Jakarta EE 10 beta feature, available in Open Liberty 22.0.0.3-beta.  Concurrency 3.0 introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.

Author: ```suggestion
Although only a few short months have passed since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, we're already implementing Jakarta EE 10 features. Concurrency 3.0 marks our first Jakarta EE 10 beta feature, available in Open Liberty 22.0.0.3-beta. Concurrency 3.0 introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
```

== All Beta Features package
The All Beta features package for link:{url-about}[Open Liberty] contains all Open Liberty beta features and GA features and functions.  This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.
If you try the beta package, <<feedback, let us know what you think>>.

The All Beta Features package includes the following beta features:

* <<concurrency, Concurrency 3.0 for Jakarta EE 10>>

[#concurrency]
=== Concurrency 3.0 for Jakarta EE 10

The Jakarta Concurrency specification enhances the set of parallel and reactive programming models available to applications, enabling tasks and completion stage actions to run with the thread context of the submitter.

> Concurrency 3.0 introduces a variety of enhancements, including...

Reviewer: Concurrency 3.0 introduces a variety of enhancements, including...
--->
Concurrency 3.0 includes the following enhancements:

> Concurrency 3.0 introduces a variety of enhancements, including...

Author: ```suggestion
Concurrency 3.0 includes the following enhancements:
```

* Resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires.
* Context-aware completion stages that are managed by the container.
* Asynchronous methods that are backed by context-aware completion stages.
* Improvements to Triggers for customizing the scheduling of business logic according to dates and times within one's own time zone.
* Standardized integration for third-party providers of thread context.

> While in beta, Open Liberty allows this Jakarta EE 10 feature to be used alongside Jakarta EE 9 features so that you can try it out with existing applications. Enable the `concurrent-3.0` feature in your `server.xml`,

Reviewer: While in beta, Open Liberty allows this Jakarta EE 10 feature to be used
--->
While in beta, Open Liberty allows you to use this Jakarta EE 10 feature...

        }
    }

    ...
----

When the application invokes the asynchronous method, the container intercepts the invocation and hands back a completion stage representing its eventual completion, which is arranged by the `ManagedExecutorService`.

[source, java]
----
    public void doGet(HttpServletRequest req, HttpServletResponse res) {
        ...
        accounts.findOverdue(60).thenAccept(this::sendReminder);
        ...
    }

    void sendReminder(List<Account> overdue) {
        try {
            // Resource reference lookup requires context of the application component,
            DataSource db = InitialContext.doLookup("java:comp/env/jdbc/customer-db-ref");
            ...
        } catch (NamingException | SQLException x) {
            throw new CompletionException(x);
        }
    }
----

The Concurrency 3.0 specification hasn't been published yet because Jakarta EE 10 is still under development. However, you can preview a draft of it at the link:https://github.com/eclipse-ee4j/concurrency-api/blob/2abe74ce30589ef53e248009fcb446491fd42a27/specification/src/main/asciidoc/jakarta-concurrency.adoc[concurrency-api GitHub page].

> Similarly, the API jar will not be available on Maven until the specification is released, but in the mean time, you can check out the link:http://public.dhe.ibm.com/ibmdl/export/pub/software/olrepo/io/openliberty/jakarta/enterprise/concurrent/jakarta.enterprise.concurrent-api/3.0.0.20220120/[draft copy.]

Reviewer: mean time
--->
meantime

> Similarly, the API jar will not be available on Maven until the specification is released, but in the mean time, you can check out the link:http://public.dhe.ibm.com/ibmdl/export/pub/software/olrepo/io/openliberty/jakarta/enterprise/concurrent/jakarta.enterprise.concurrent-api/3.0.0.20220120/[draft copy.]

Author: ```suggestion
Similarly, the API jar will not be available on Maven until the specification is released, but in the meantime, you can check out the link:http://public.dhe.ibm.com/ibmdl/export/pub/software/olrepo/io/openliberty/jakarta/enterprise/concurrent/jakarta.enterprise.concurrent-api/3.0.0.20220120/[draft copy.]
```

author_github: https://github.com/ryan-storey
seo-title: Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta - OpenLiberty.io
seo-description: Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
blog_description: Open Liberty 22.0.0.3-beta offers Concurrency 3.0 for Jakarta EE 10, which introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Concurrency 3.0 for Jakarta EE 10 in Open Liberty 22.0.0.3-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> It's only been a couple of short months since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, and we're already looking forward as we collaborate on and begin to implement Jakarta EE 10 features.  Open Liberty 22.0.0.3-beta offers our first Jakarta EE 10 beta feature, Concurrency 3.0.  Concurrency 3.0 introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.

Author: ```suggestion
It's only been a few short months since we announced Jakarta EE 9.1 support in Open Liberty 21.0.0.12, and we're already looking forward as we collaborate on and begin to implement Jakarta EE 10 features.  Concurrency 3.0 marks our first Jakarta EE 10 beta feature, available in Open Liberty 22.0.0.3-beta.  Concurrency 3.0 introduces a variety of new enhancements including resource definition annotations that put applications in control of defining the concurrency constraints and thread context types that the application requires, context-aware completion stages that are managed by the container, and more.
```

== All Beta Features package

> The All Beta features package for link:{url-about}[Open Liberty] contains all Open Liberty beta features and GA features and functions.
> 
> This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.
> 
> If you try the beta package, <<feedback, let us know what you think>>.

Author: ```suggestion
The All Beta features package for link:{url-about}[Open Liberty] contains all Open Liberty beta features and GA features and functions.  This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.
If you try the beta package, <<feedback, let us know what you think>>.
```
